### PR TITLE
parser: support paths in bang macro invocations (e.g. `path::to::macro!()`)

### PR DESCRIPTION
**Pull request by jseyfried**
Groundwork for macro modularization (cc #35896).
r? @nrc

### CODE REVIEW

File: src/test/compile-fail/macro-context.rs [hunk 1]
```diff
@@ -14,11 +14,8 @@ macro_rules! m {
                             //~| ERROR macro expansion ignores token `typeof`
                             //~| ERROR macro expansion ignores token `;`
                             //~| ERROR macro expansion ignores token `;`
-                            //~| ERROR macro expansion ignores token `i`
 }
 
-m!();               //~ NOTE the usage of `m!` is likely invalid in item context
-
```
Comment (jseyfried) [inline src/test/compile-fail/macro-context.rs]: After this PR, this triggers a parse error on expansion (``expected one of `!` or `::` ``).

**Comment by jseyfried** [comment]
cc @petrochenkov

**Comment by petrochenkov** [comment]
Looks good.

Note to myself: audit all places where disambiguation "path vs ident" happens, there are still some inconsistencies.

**Comment by nrc** [comment]
Am I right that this PR means path macros can be parsed but will always give an error later on before being expanded? If not, then it'd be good to have a feature gate in here.

Could you add some tests too please?

**Comment by jseyfried** [comment]
Yeah, `expand_bang_invoc` always errors with non-trivial paths -- I'll added tests.

**Comment by jseyfried** [comment]
Added tests.

**Pull request merged**